### PR TITLE
Feat/authentication

### PR DESCRIPTION
**Pull request by hnimtadd**
added authentication packages with JWT token

### CODE REVIEW

File: functions/authentication/go.mod [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module blinder/functions/authentication
```
Comment (tanlethanh) [inline functions/authentication/go.mod:1]: use `blinders/functions/auth`

File: functions/authentication/main.go
Comment (tanlethanh) [inline functions/authentication/main.go:1]: we just use `auth` as a package

File: go.work [hunk 2]
```diff
@@ -1,10 +1,13 @@
 go 1.21.3
 
 use (
+	./functions/authentication
```
Comment (tanlethanh) [inline go.work:4]: not use auth lambda function

File: go.work [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
 	./functions/collecting
 	./functions/flashcard
 	./functions/suggestion
 	./functions/translation
+	./packages/authentication
 	./packages/suggestion
 	./packages/translation
+	./services/authentication
```
Comment (tanlethanh) [inline go.work:12]: I prefer use `auth` for short name convention in Go

File: packages/authentication/Makefile
Comment (tanlethanh) [inline packages/authentication/Makefile:1]: I don't think we actually need this `Makefile`

File: packages/authentication/.gitignore
Comment (tanlethanh) [inline packages/authentication/.gitignore:1]: just use `.gitignore` at project scope

File: packages/authentication/firebase_auth_maker.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+type FirebaseAuthManager struct {
+	authClient *auth.Client
+}
+
+func NewFirebaseAuthManager() (Maker, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	// init firebase admin app with GOOGLE_APPLICATION_CREDENTIALS variable
```
Comment (tanlethanh) [inline packages/authentication/firebase_auth_maker.go:19]: remove this comment, the code is understandable

File: packages/authentication/firebase_auth_maker.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+func (m *FirebaseAuthManager) Generate(user *User) (string, error) {
+	// Generate token are generated with firebase backend server
+	// ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	// defer cancel()
+	// token, err := m.authClient.GetUserByProviderID(ctx, user.ID)
+	//
+	// if err != nil {
+	// 	return "", err
+	// }
+	// return token, nil
```
Comment (tanlethanh) [inline packages/authentication/firebase_auth_maker.go:45]: same as above comment

File: packages/authentication/firebase_auth_maker.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+		return nil, err
+	}
+
+	manager := &FirebaseAuthManager{
+		authClient: authClient,
+	}
+	return manager, nil
+}
+
+func (m *FirebaseAuthManager) Generate(user *User) (string, error) {
```
Comment (tanlethanh) [inline packages/authentication/firebase_auth_maker.go:36]: I think method of a struct should be put below, following the definition of struct

Comment (tanlethanh) [reply packages/authentication/firebase_auth_maker.go:36]: also just using pointer when we need to mutate the object

File: packages/authentication/firebase_auth_maker.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+import (
+	"context"
+	"time"
+
+	firebase "firebase.google.com/go/v4"
+	"firebase.google.com/go/v4/auth"
+)
+
+type FirebaseAuthManager struct {
```
Comment (tanlethanh) [inline packages/authentication/firebase_auth_maker.go:11]: I prefer naming it `FirebaseAuthManager` implements `AuthManager`

File: packages/authentication/firebase_auth_maker.go
Comment (tanlethanh) [inline packages/authentication/firebase_auth_maker.go:1]: Prefer using `camelCase` name convention for Go

Comment (tanlethanh) [reply packages/authentication/firebase_auth_maker.go:1]: Prefer naming this file `firebase.go`

File: packages/authentication/firebase_auth_maker_test.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+	assert.Nil(t, err)
+	assert.NotNil(t, maker)
+	var (
+		validUser = &User{
+			ID:    "t7ZYtyjYCbMxOefUALu8b2P4AVO2",
+			Email: "minhdat15012002@gmail.com",
+		}
+	)
+	validToken, err := GetUserIDToken(validUser)
+	// validToken, err := maker.Generate(validUser)
```
Comment (tanlethanh) [inline packages/authentication/firebase_auth_maker_test.go:37]: remove this line

File: packages/authentication/firebase_auth_maker.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+	authClient *auth.Client
+}
+
+func NewFirebaseAuthManager() (Maker, error) {
```
Comment (tanlethanh) [inline packages/authentication/firebase_auth_maker.go:15]: should pass the path of credentials file as a parameter

File: packages/authentication/firebase_auth_maker_test.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+	assert.Nil(t, err)
+	assert.NotEmpty(t, validToken)
+	u, err := maker.Verify(validToken)
+	assert.Nil(t, err)
+	assert.NotNil(t, u)
+	assert.Equal(t, u, validUser)
+}
```
Comment (tanlethanh) [inline packages/authentication/firebase_auth_maker_test.go:44]: should have space between function definitions

File: packages/authentication/firebase_auth_maker_test.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+	maker, err := NewFirebaseAuthManager()
+	assert.Nil(t, err)
+	assert.NotNil(t, maker)
+	var (
+		invalidToken = "invalidTOken"
+	)
+	u, err := maker.Verify(invalidToken)
+
+	// validToken, err := maker.Generate(validUser)
```
Comment (tanlethanh) [inline packages/authentication/firebase_auth_maker_test.go:62]: remove this line

File: packages/authentication/go.mod [hunk 12]
```diff
@@ -0,0 +1,44 @@
+module blinders/packages/authentication
+
+go 1.21.3
+
+require (
+	firebase.google.com/go/v4 v4.12.1
+	github.com/golang-jwt/jwt/v5 v5.1.0
```
Comment (tanlethanh) [inline packages/authentication/go.mod:7]: remove the dependency we not use

File: packages/authentication/firebase_auth_maker_test.go [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+		return "", err
+	}
+	return idToken, nil
+}
+
+const (
+	verifyCustomTokenURL = "https://www.googleapis.com/identitytoolkit/v3/relyingparty/verifyCustomToken?key=%s"
+)
+
+var (
```
Comment (tanlethanh) [inline packages/authentication/firebase_auth_maker_test.go:93]: use const instead

File: packages/authentication/jwt_maker.go
Comment (tanlethanh) [inline packages/authentication/jwt_maker.go:1]: remove this file

File: packages/authentication/jwt_maker_test.go
Comment (tanlethanh) [inline packages/authentication/jwt_maker_test.go:1]: remove this file

File: packages/authentication/main.go
Comment (tanlethanh) [inline packages/authentication/main.go:1]: remove this file

File: packages/authentication/maker.go [hunk 14]
```diff
@@ -0,0 +1,6 @@
+package authentication
+
+type Maker interface {
+	Generate(user *User) (string, error)
+	Verify(token string) (*User, error)
+}
```
Comment (tanlethanh) [inline packages/authentication/maker.go:6]: should be `AuthManager`

File: packages/authentication/user.go [hunk 15]
```diff
@@ -0,0 +1,7 @@
+package authentication
+
+// User type in Authenticate scope, maybe use User type in common package
+type User struct {
+	ID    string `json:"user_id"`    // firebase id of user
+	Email string `json:"user_email"` // gmail address of user
```
Comment (tanlethanh) [inline packages/authentication/user.go:6]: with JSON, prefer using `camelCase`

File: packages/authentication/user.go [hunk 16]
```diff
@@ -0,0 +1,7 @@
+package authentication
+
+// User type in Authenticate scope, maybe use User type in common package
```
Comment (tanlethanh) [inline packages/authentication/user.go:3]: remove this comment

**Comment by hnimtadd** [comment]
@tanlethanh, I already made some changes, you could review it. The code is tested.

**Review by tanlethanh — APPROVED** [review]

**Pull request merged**